### ISSUE TITLE
Spike: Figure out what e2e-deploy, etc. need for us to use tags

### ISSUE THREAD

**Issue by katherine-black** [cloudigrade/cloudigrade#715]
In GitLab by @infinitewarp on Jul 15, 2020, 09:22

## Summary

We need e2e-deploy (and friends) to use git tags to drive the process that creates images and deploys them to various environments. We need to understand what configuration changes we need to make in order for that to happen.

Expected outcome:

- [ ] Shared documentation/notes from a conversation with e2e-deploy maintainers that explains how we need to reconfigure e2e-deploy for our needs.
- [ ] Shared documentation/notes about what exactly e2e-deploy will do to build images and deploy to each environment (CI, QA, prod (old OSDv3), stage (new OSDv4), prod (new OSDv4)) when we define new git tags.

Time box: 2 days

## Assumptions and Questions

- If discussions with e2e-deploy maintainers suggests this is not possible, we assert that it needs to be.
- Possible implementation idea for e2e-deploy: Can we dictate through some webhook to e2e/etc. what specific tag to use? What specific image to use?
- "Use a stable branch/tag" is not a good enough solution. See our previous discussions about a release process.
- QE folks will drive this conversation, but feel free to lean on the devs if you have questions or want feedback along the way.
- **IMPORTANT**: With the new OSDv4 production cluster rollout, there is a whole new layer of abstraction to deal with. Please remember to include understanding how AppSRE's process around OSDv4 might affect any plans for tagging images here.
  - [saas-templates](https://gitlab.cee.redhat.com/insights-platform/saas-templates/) effectively gets a modified copy of the contents of [e2e-deploy](https://github.com/RedHatInsights/e2e-deploy/), and that is referenced by [app-interface](https://gitlab.cee.redhat.com/service/app-interface) for the actual deployment process.
- For general reference, though this shouldn't be necessary to have for this research spike, here is where each of our deployments live:
  - ci: https://console.insights-dev.openshift.com/console/project/cloudigrade-ci/overview
  - qa: https://console.insights-dev.openshift.com/console/project/cloudigrade-qa/overview
  - prod (OSDv3): https://console.insights.openshift.com/console/project/cloudigrade-prod/overview
  - stage (OSDv4): https://console-openshift-console.apps.crc-stg-01.o4v9.p1.openshiftapps.com/k8s/ns/cloudigrade-stage/deployments
  - prod (OSDv4): https://console-openshift-console.apps.crcp01ue1.o9m8.p1.openshiftapps.com/k8s/cluster/projects/cloudigrade-prod

**Comment by katherine-black** [comment]
In GitLab by @mpierce on Sep 24, 2020, 17:01

Notes from [tech talk](https://drive.google.com/file/d/1R3VvB8VXP8cUkWrVPyaHcf6PSimFYF7A/view) re: testing in the app-sre pipeline:

tl;dr
The strategy will create ephemeral environments with automation that watches for changes in PRs and spins up new environments. Seems close to what we want to have except that there seems to be a set number of ephemeral environments available rather than creating them on the fly for specific sets of changes.
These strategies are presented as one possible strategy and a request for feedback, so it seems like a reasonable time to offer suggestions to include resources to meet our needs.

What currently exists:

- Smoke tests and how they work (the old way)
    - A jenkins job runs against a PR
      - looks up github merge commit ref for a particular PR
      - deploys custom build config for the app in question pointing at the PR branch (e2e-deploy)
      - deploys app and dependencies into namespace (using `e2e-deploy` and `-e smoke`
      - spins up iqe test pod
      - runs all iqe tests with specified marker pointing at ephemeral environment

What is coming:

- no more e2e-deploy, apps now configured via app-interface (single source of truth re: how apps are configured)

1. Dev makes changes to code
1. Dev creates PR in Github
1. PR fires webhooks check_pr script (jenkins)
1. This invokes Bonfire system which goes to get deployment templates from App-Interface
1. Deploys those templates into ephemeral namaspace
1. Clowder operator takes those templates and deploys the necessary applications and all of the base resources needed
1. Once all is deployed, iqe test suite gets run against it

**[Bonfire](https://github.com/RedHatInsights/bonfire)** 

POC example using Ingress service

See also: [Bonfire/Clowder intro](https://primetime.bluejeans.com/a2m/events/playback/7b915a4f-2481-449e-9437-902b2ae35834)

1. There is an app config in app-interface called *Ephemeral* which defines all the ephemeral namespaces
    - So there's a pool of ephemeral namespaces usable by apps for this testing (currently there's just one).
    - There's one called *ephemeral base* where secrets are defined for the namespaces - App-sre pulls the secrets from vault and inserts them into the namespace (all secrets in one place for all namespaces). *ephemeral base* also used for the applications deployment config (see Ingress deployment file)
1. Bonfire is the thing that lets application config spin up ephemeral environments (currently all this is done via the `pr_check.sh` script)
    - check in or out a namespace
    - copy secrets from the base namespace into whatever one you need them in
    - get Kubernetes config for an application
    - list namespaces that are available
    - wait for all necessary resources to be ready before starting
    - in the future - `deploy` command will tie all these things together
    - also in the future - some flag or tag on a PR that defines if smoke tests are actually needed to avoid unnecessary runs of the tests.

**[Clowder](https://github.com/RedHatInsights/clowder)**
Resources deployed for Clowder to work against
    - base infrastructure resources necessary
    - app what is needed for the specific app in question

See also: [Clowder/Bonfire introduction](https://primetime.bluejeans.com/a2m/events/playback/7b915a4f-2481-449e-9437-902b2ae35834)
